README.rst
===============================================
isis2json: CDS/ISIS to JSON database converter
===============================================

The isis2json.py is a Python/Jython script to export ISIS (MST+XRF)
or ISO-2709 databases to JSON files, optionally compatible with CouchDB
and MongoDB.

Running under Jython, both MST+XRF and ISO-2709 files can be read,
thanks to the Bruma Java library found in the lib/ directory.

Running under Python, only ISO-2709 files can be read.

Usage
======

::

  $ ./isis2json.py -h
  usage: isis2json.py [-h] [-o OUTPUT.json] [-c] [-m] [-t ISIS_JSON_TYPE]
                      [-q QTY] [-s SKIP] [-i TAG_NUMBER] [-u] [-p PREFIX]
                      [-n] [-k TAG:VALUE]
                      INPUT.(mst|iso)

  Convert an ISIS .mst or .iso file to a JSON array

  positional arguments:
    INPUT.(mst|iso)     .mst or .iso file to read

  optional arguments:
    -h, --help          show this help message and exit
    -o OUTPUT.json, --out OUTPUT.json
                        the file where the JSON output should be written
                          (default: write to stdout)
    -c, --couch         output array within a "docs" item in a JSON document
                          for bulk insert to CouchDB via POST to db/_bulk_docs
    -m, --mongo         output individual records as separate JSON objects,
                          one per line for bulk insert to MongoDB via
                          mongoimport utility
    -t ISIS_JSON_TYPE, --type ISIS_JSON_TYPE
                        ISIS-JSON type, sets field structure:
                          1=string, 2=alist, 3=dict
    -q QTY, --qty QTY   maximum quantity of records to read (default=ALL)
    -s SKIP, --skip SKIP  records to skip from start of .mst (default=0)
    -i TAG_NUMBER, --id TAG_NUMBER
                        generate an "_id" from the given unique TAG field
                          number for each record
    -u, --uuid          generate an "_id" with a random UUID for each record
    -p PREFIX, --prefix PREFIX
                        concatenate prefix to every numeric field tag
                          (ex. 99 becomes "v99")
    -n, --mfn           generate an "_id" from the MFN of each record
                          (available only for .mst input)
    -k TAG:VALUE, --constant TAG:VALUE
                        Include a constant tag:value in every record
                          (ex. -k type:AS)


ISIS-JSON Record Types
=======================

There are many ways to represent CDS/ISIS records in JSON [#]_. This
utility currently exports ISIS-JSON types 1, 2 and 3.

Given an ISIS record with this strcuture::

   2 «538886»
  10 «Kanda, Paulo Afonso^1USP^2FMUSP^3CRDC^pBrasil^cSão Paulo^rorg»
  10 «Smidth, Magali Taino^1USP^2FMUSP^3CRDC^pBrasil^cSão Paulo^rorg»

Below are the three supported representations of that record in JSON:

ISIS-JSON type 1
-----------------

::

  {"10":
      ["Kanda, Paulo Afonso^1USP^2FMUSP^3CRDC^pBrasil^cSão Paulo^rorg",
       "Smidth, Magali Taino^1USP^2FMUSP^3CRDC^pBrasil^cSão Paulo^rorg"],
   "2":
      ["538886"]
  }

ISIS-JSON type 2
-----------------

::

    {"10":
        [
            [
                ("_", "Kanda, Paulo Afonso"),
                ("1", "USP"),
                ("2", "FMUSP"),
                ("3", "CRDC"),
                ("p", "Brasil"),
                ("c", "São Paulo"),
                ("r", "org")
            ],
            [
                ("_", "Smidth, Magali Taino"),
                ("1", "USP"),
                ("2", "FMUSP"),
                ("3", "CRDC"),
                ("p", "Brasil"),
                ("c", "São Paulo"),
                ("r", "org")
            ]
        ],
     "2":
        [
            [
                ("_", "538886")
            ]
        ]
    }

ISIS-JSON type 3
-----------------

::

    {"10":
        [
            {
                "_": "Kanda, Paulo Afonso",
                "1": "USP",
                "2": "FMUSP",
                "3": "CRDC",
                "c": "São Paulo",
                "p": "Brasil",
                "r": "org"
            },
            {
                "_": "Smidth, Magali Taino",
                "1": "USP",
                "2": "FMUSP",
                "3": "CRDC",
                "c": "São Paulo",
                "p": "Brasil",
                "r": "org"
            }
        ],
     "2":
        [
            {
                "_": "538886"
            }
        ]
    }


.. [#] See section 4.1 of http://journal.code4lib.org/articles/4893


Dependencies
=============

Under Python, isis2json.py depends on:

- Python2.6 or 2.7
- argparse.py (bundled; also part of the CPython 2.7 distribution)

Under Jython, isis2json.py depends on:

- Jython 2.5;
- argparse.py (bundled)
- Bruma.jar on the CLASSPATH (bundled);
- jyson-1.0.1.jar on the CLASSPATH (bundled);

Example CLASSPATH::

  export CLASSPATH=/home/luciano/lib/Bruma.jar:/home/luciano/lib/jyson-1.0.1.jar


Troubleshooting
================

SyntaxError on `yield fields` running isis2json.py under Jython
-------------------------------------------------------------------

If you see this::

  Traceback (innermost last):
    (no code object) at line 0
    File "./isis2json.py", line 84
          yield fields
              ^
  SyntaxError: invalid syntax

You are probably running Jython 2.2, an old version that is packaged
with several Linux distributions such as Debian and Ubuntu. To verify,
type::

  $ jython --version
  Jython 2.2.1 on java1.6.0_20

To fix, download and install Jython 2.5 or later from the Jython project
on SourceForge_.

.. _SourceForge: http://sourceforge.net/projects/jython/files/jython/

IMPORT ERROR: Jython 2.5 and Bruma.jar are required to read .mst files
-----------------------------------------------------------------------

Check if Jython 2.5 or later is installed::

  $ jython --version
  Jython 2.5.2

If it is not, se issue above. If it is, add the path to Bruma.jar to
the CLASSPATH environment variable, or pass it via the `jython -J-cp`
command line option when running isis2json.py, like this::

  $ jython -J-cp lib/jyson-1.0.1.jar:lib/Bruma.jar isis2json.py fixtures/LILACS1.mst




HISTORY
The code in this directory was moved from an SVN repository [1] to allow
easier collaboration.

[1] http://svn.reddes.bvsalud.org/isisnbp/tools

Later, the isis2json codebase was separated from the ISIS-NBP project
to make it easier to install and use.

The canonical, most up-to-date version of these files is at this GitHub
isis2json project.

For historic purposes, below is the output of a some commands showing
the files in the original repository.


#####################################################################

luciano@bireme:~/prj/isisnbp/svn/tools$ ls -la
total 217
drwxrwxrwx 1 root root  4096 2011-02-07 16:23 .
drwxrwxrwx 1 root root  4096 2010-12-20 18:47 ..
-rwxrwxrwx 1 root root 87352 2010-11-30 17:00 argparse.py
-rwxrwxrwx 1 root root   365 2010-11-30 17:00 bulkup.sh
-rwxrwxrwx 1 root root 18092 2010-11-30 17:00 COPYING
-rwxrwxrwx 1 root root 18092 2010-11-30 17:00 COPYING.LESSER
-rwxrwxrwx 1 root root  2522 2010-11-30 17:00 idfile.py
-rwxrwxrwx 1 root root  4839 2010-11-30 17:00 idfile_test.txt
-rwxrwxrwx 1 root root  9551 2010-12-20 18:47 isis2json.py
-rwxrwxrwx 1 root root  5391 2010-11-30 17:00 iso2709.py
-rwxrwxrwx 1 root root  3766 2010-11-30 17:00 iso2709_test.txt
-rwxrwxrwx 1 root root  6072 2011-02-07 16:23 lixo
-rwxrwxrwx 1 root root   415 2010-11-30 17:00 README.txt
-rwxrwxrwx 1 root root  8413 2010-11-30 17:00 schematize.py
-rwxrwxrwx 1 root root  5593 2010-11-30 17:00 schematize_test.txt
-rwxrwxrwx 1 root root  5759 2010-11-30 17:00 subfield_access.py
-rwxrwxrwx 1 root root  3087 2010-11-30 17:00 subfield.py
drwxrwxrwx 1 root root  4096 2011-02-07 16:20 .svn

#####################################################################

luciano@bireme:~/prj/isisnbp/svn/tools$ svn info
Caminho: .
URL: http://svn.reddes.bvsalud.org/isisnbp/tools
Raiz do Repositório: http://svn.reddes.bvsalud.org/isisnbp
UUID do repositório: 7b3c32f5-e272-4d6f-9936-548b95b19007
Revisão: 2513
Tipo de Nó: diretório
Agendado: normal
Autor da Última Mudança: fabio.batalha
Revisão da Última Mudança: 2504
Data da Última Mudança: 2010-12-13 17:25:53 -0200 (Seg, 13 Dez 2010)

#####################################################################

luciano@bireme:~/prj/isisnbp/svn/tools$ svn log

------------------------------------------------------------------------
r2504 | fabio.batalha | 2010-12-13 17:25:53 -0200 (Seg, 13 Dez 2010) | 1 linha

Including parameter regtype to include a field for each register that could discribe the register type
------------------------------------------------------------------------
r2489 | luciano.ramalho | 2010-11-13 21:18:13 -0200 (Sáb, 13 Nov 2010) | 1 linha

renamed subfield canonical subfield parser module
------------------------------------------------------------------------
r2488 | luciano.ramalho | 2010-11-13 21:16:52 -0200 (Sáb, 13 Nov 2010) | 1 linha

implemented new subfield parser with code options, renamed old subfield access tests
------------------------------------------------------------------------
r2483 | luciano.ramalho | 2010-11-01 09:07:56 -0200 (Seg, 01 Nov 2010) | 1 linha

added option to add prefix to numeric tags
------------------------------------------------------------------------
r2473 | luciano.ramalho | 2010-10-25 00:37:44 -0200 (Seg, 25 Out 2010) | 1 linha

new, more robust implementation of subfield parser
------------------------------------------------------------------------
r2466 | luciano.ramalho | 2010-10-17 14:14:29 -0200 (Dom, 17 Out 2010) | 1 linha

fixed bug when generating _id for non-expanded JSON
------------------------------------------------------------------------
r2459 | luciano.ramalho | 2010-10-12 03:20:06 -0300 (Ter, 12 Out 2010) | 1 linha

expand composite field into object (dict)
------------------------------------------------------------------------
r2446 | luciano.ramalho | 2010-09-27 18:55:01 -0300 (Seg, 27 Set 2010) | 1 linha

added LGPL headers and files
------------------------------------------------------------------------
r2444 | luciano.ramalho | 2010-09-27 17:52:08 -0300 (Seg, 27 Set 2010) | 1 linha

added LGPL header and files
------------------------------------------------------------------------
r2443 | luciano.ramalho | 2010-09-27 10:57:03 -0300 (Seg, 27 Set 2010) | 1 linha

added quorum option to ignore infrequent tags
------------------------------------------------------------------------
r2442 | luciano.ramalho | 2010-09-20 21:55:14 -0300 (Seg, 20 Set 2010) | 1 linha

added unique key check
------------------------------------------------------------------------
r2441 | luciano.ramalho | 2010-09-20 00:16:01 -0300 (Seg, 20 Set 2010) | 1 linha

first version of schematize.py
------------------------------------------------------------------------
r2437 | luciano.ramalho | 2010-09-19 18:16:06 -0300 (Dom, 19 Set 2010) | 1 linha

started schematize script
------------------------------------------------------------------------
r2436 | luciano.ramalho | 2010-09-18 18:28:11 -0300 (Sáb, 18 Set 2010) | 1 linha

added subield.py with comparing subfield handling with and without regexes; minor tweak to isis2json Zeus output: when SKIP_INACTIVE is true, do not add 'active' field to ouptut, as it is redundant
------------------------------------------------------------------------
r2435 | luciano.ramalho | 2010-09-17 19:51:50 -0300 (Sex, 17 Set 2010) | 1 linha

added tests for a few corner cases
------------------------------------------------------------------------
r2434 | luciano.ramalho | 2010-09-17 19:33:10 -0300 (Sex, 17 Set 2010) | 1 linha

created id file reader
------------------------------------------------------------------------
r2433 | luciano.ramalho | 2010-08-30 07:38:51 -0300 (Seg, 30 Ago 2010) | 1 linha

updated bulkup.sh with new couch option
------------------------------------------------------------------------
r2432 | luciano.ramalho | 2010-08-30 00:45:52 -0300 (Seg, 30 Ago 2010) | 1 linha

fixed error messages in writeJsonArray which assumed availability of an mfn

test.sh
#!/bin/bash

PYTHONPATH=../isisdm/:$PYTHONPATH

python isis2json.py -t $1 ../fixtures/lilacs1/LILACS.iso

#jython isis2json.py -t $1 ../fixtures/lilacs1/LILACS.mst


isis2json.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

# isis2json.py: convert ISIS and ISO-2709 files to JSON
#
# Copyright (C) 2010 BIREME/PAHO/WHO
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published
# by the Free Software Foundation, either version 2.1 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.

# You should have received a copy of the GNU Lesser General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

############################
# this script works with Python or Jython (versions >=2.5 and <3)

import sys
import argparse
from uuid import uuid4
import os

try:
    import json
except ImportError:
    if os.name == 'java': # running Jython
        from com.xhaus.jyson import JysonCodec as json
    else:
        import simplejson as json

SKIP_INACTIVE = True
DEFAULT_QTY = 2**31
ISIS_MFN_KEY = 'mfn'
ISIS_ACTIVE_KEY = 'active'
SUBFIELD_DELIMITER = '^'
INPUT_ENCODING = 'cp1252'

def iterMstRecords(master_file_name, isis_json_type):
    try:
        from bruma.master import MasterFactory, Record
    except ImportError:
        print('IMPORT ERROR: Jython 2.5 and Bruma.jar are required to read .mst files')
        raise SystemExit
    mst = MasterFactory.getInstance(master_file_name).open()
    for record in mst:
        fields = {}
        if SKIP_INACTIVE:
            if record.getStatus() != Record.Status.ACTIVE:
                continue
        else: # save status only there are non-active records
            fields[ISIS_ACTIVE_KEY] = record.getStatus() == Record.Status.ACTIVE
        fields[ISIS_MFN_KEY] = record.getMfn()
        for field in record.getFields():
            field_key = str(field.getId())
            field_occurrences = fields.setdefault(field_key,[])
            if isis_json_type == 3:
                content = {}
                for subfield in field.getSubfields():
                    subfield_key = subfield.getId()
                    if subfield_key == '*':
                        content['_'] = subfield.getContent()
                    else:
                        subfield_occurrences = content.setdefault(subfield_key,[])
                        subfield_occurrences.append(subfield.getContent())
                field_occurrences.append(content)
            elif isis_json_type == 1:
                content = []
                for subfield in field.getSubfields():
                    subfield_key = subfield.getId()
                    if subfield_key == '*':
                        content.insert(0, subfield.getContent())
                    else:
                        content.append(SUBFIELD_DELIMITER+subfield_key+
                                       subfield.getContent())
                field_occurrences.append(''.join(content))
            else:
                raise NotImplementedError('ISIS-JSON type %s conversion not yet implemented for .mst input' % isis_json_type)
        yield fields
    mst.close()

def iterIsoRecords(iso_file_name, isis_json_type):
    from iso2709 import IsoFile
    from subfield import expand

    iso = IsoFile(iso_file_name)
    for record in iso:
        fields = {}
        for field in record.directory:
            field_key = str(int(field.tag)) # remove leading zeroes
            field_occurrences = fields.setdefault(field_key,[])
            content = field.value.decode(INPUT_ENCODING,'replace')
            if isis_json_type == 1:
                field_occurrences.append(content)
            elif isis_json_type == 2:
                field_occurrences.append(expand(content))
            elif isis_json_type == 3:
                field_occurrences.append(dict(expand(content)))
            else:
                raise NotImplementedError('ISIS-JSON type %s conversion not yet implemented for .iso input' % isis_json_type)

        yield fields
    iso.close()

def writeJsonArray(iterRecords, file_name, output, qty, skip, id_tag,
                   gen_uuid, mongo, mfn, isis_json_type, prefix, constant):
    start = skip
    end = start + qty
    if not mongo:
        output.write('[')
    if id_tag:
        id_tag = str(id_tag)
        ids = set()
    else:
        id_tag = ''
    for i, record in enumerate(iterRecords(file_name, isis_json_type)):
        if i >= end:
            break
        if i > start and not mongo:
            output.write(',')
        output.write('\n')
        if start <= i < end:
            if id_tag:
                occurrences = record.get(id_tag, None)
                if occurrences is None:
                    msg = 'id tag #%s not found in record %s'
                    if ISIS_MFN_KEY in record:
                        msg = msg + (' (mfn=%s)' % record[ISIS_MFN_KEY])
                    raise KeyError(msg % (id_tag, i))
                if len(occurrences) > 1:
                    msg = 'multiple id tags #%s found in record %s'
                    if ISIS_MFN_KEY in record:
                        msg = msg + (' (mfn=%s)' % record[ISIS_MFN_KEY])
                    raise TypeError(msg % (id_tag, i))
                else: # ok, we have one and only one id field
                    if isis_json_type == 1:
                        id = occurrences[0]
                    elif isis_json_type == 2:
                        id = occurrences[0][0][1]
                    elif isis_json_type == 3:
                        id = occurrences[0]['_']
                    if id in ids:
                        msg = 'duplicate id %s in tag #%s, record %s'
                        if ISIS_MFN_KEY in record:
                            msg = msg + (' (mfn=%s)' % record[ISIS_MFN_KEY])
                        raise TypeError(msg % (id, id_tag, i))
                    record['_id'] = id
                    ids.add(id)
            elif gen_uuid:
                record['_id'] = unicode(uuid4())
            elif mfn:
                record['_id'] = record[ISIS_MFN_KEY]
            if prefix:
                for tag in tuple(record): # iterate over a fixed sequence of tags
                    if str(tag).isdigit():
                        record[prefix+tag] = record[tag]
                        del record[tag] # this is why we iterate over a tuple
                        # with the tags, and not directly on the record dict
            if constant:
                constant_key, constant_value = constant.split(':')
                record[constant_key] = constant_value
            output.write(json.dumps(record).encode('utf-8'))
    if not mongo:
        output.write('\n]')
    output.write('\n')

if __name__ == '__main__':

    # create the parser
    parser = argparse.ArgumentParser(
        description='Convert an ISIS .mst or .iso file to a JSON array')

    # add the arguments
    parser.add_argument(
        'file_name', metavar='INPUT.(mst|iso)', help='.mst or .iso file to read')
    parser.add_argument(
        '-o', '--out', type=argparse.FileType('w'), default=sys.stdout,
        metavar='OUTPUT.json',
        help='the file where the JSON output should be written'
             ' (default: write to stdout)')
    parser.add_argument(
        '-c', '--couch', action='store_true',
        help='output array within a "docs" item in a JSON document'
             ' for bulk insert to CouchDB via POST to db/_bulk_docs')
    parser.add_argument(
        '-m', '--mongo', action='store_true',
        help='output individual records as separate JSON dictionaries,'
             ' one per line for bulk insert to MongoDB via mongoimport utility')
    parser.add_argument(
        '-t', '--type', type=int, metavar='ISIS_JSON_TYPE', default=1,
        help='ISIS-JSON type, sets field structure: 1=string, 2=alist, 3=dict')
    parser.add_argument(
        '-q', '--qty', type=int, default=DEFAULT_QTY,
        help='maximum quantity of records to read (default=ALL)')
    parser.add_argument(
        '-s', '--skip', type=int, default=0,
        help='records to skip from start of .mst (default=0)')
    parser.add_argument(
        '-i', '--id', type=int, metavar='TAG_NUMBER', default=0,
        help='generate an "_id" from the given unique TAG field number'
             ' for each record')
    parser.add_argument(
        '-u', '--uuid', action='store_true',
        help='generate an "_id" with a random UUID for each record')
    parser.add_argument(
        '-p', '--prefix', type=str, metavar='PREFIX', default='',
        help='concatenate prefix to every numeric field tag (ex. 99 becomes "v99")')
    parser.add_argument(
        '-n', '--mfn', action='store_true',
        help='generate an "_id" from the MFN of each record'
             ' (available only for .mst input)')
    parser.add_argument(
        '-k', '--constant', type=str, metavar='TAG:VALUE', default='',
        help='Include a constant tag:value in every record (ex. -k type:AS)')

    '''
    # TODO: implement this to export large quantities of records to CouchDB
    parser.add_argument(
        '-r', '--repeat', type=int, default=1,
        help='repeat operation, saving multiple JSON files'
             ' (default=1, use -r 0 to repeat until end of input)')
    '''
    # parse the command line
    args = parser.parse_args()
    if args.file_name.lower().endswith('.mst'):
        iterRecords = iterMstRecords
    else:
        if args.mfn:
            print('UNSUPORTED: -n/--mfn option only available for .mst input.')
            raise SystemExit
        iterRecords = iterIsoRecords
    if args.couch:
        args.out.write('{ "docs" : ')
    writeJsonArray(iterRecords, args.file_name, args.out, args.qty, args.skip,
        args.id, args.uuid, args.mongo, args.mfn, args.type, args.prefix, args.constant)
    if args.couch:
        args.out.write('}\n')
    args.out.close()

